### PR TITLE
[tests-only] Adjust pact interactions for groupTest and userTest

### PR DESCRIPTION
**Pull request by saw-jan**
part of https://github.com/owncloud/owncloud-sdk/issues/709 and https://github.com/owncloud/owncloud-sdk/issues/644

adjustment for test suites:
- groupTest
- userTest

### CODE REVIEW

File: tests/providerTest.js [hunk 1]
```diff
@@ -35,6 +36,25 @@ describe('provider testing', () => {
     }
 
     opts.stateHandlers = {
+      'user exists': (setup, params) => {
+        if (setup) {
+          fetch(process.env.PROVIDER_BASE_URL + '/ocs/v1.php/cloud/users', {
+            method: 'POST',
+            body: `userid=${params.userName}&password=${params.password}`,
```
Comment (dpakach) [inline tests/providerTest.js:43]: I think you can send simple json body here, no need to complicate this one

File: tests/providerTest.js [hunk 2]
```diff
@@ -35,6 +36,25 @@ describe('provider testing', () => {
...
+            headers: {
+              ...validAuthHeaders,
+              ...applicationFormUrlEncoded
+            }
+          })
```
Comment (dpakach) [inline tests/providerTest.js:48]: Also would be great if we check if the request actually succeded here

File: tests/providerTest.js (hunk 1 above)
Comment (saw-jan) [reply tests/providerTest.js:43]: do you mean

Comment (saw-jan) [reply tests/providerTest.js:43]: I cannot create user with json format
```json
...
body: { "userid": "user1", "password": "1234" }
headers: {
  <validauth>,
  "Content-Type": "application/json"
}
...
```

```xml
<ocs>
    <meta>
        <status>failure</status>
        <statuscode>101</statuscode>
        <message>A valid username must be provided</message>
    </meta>
    <data/>
</ocs>
```

File: tests/providerTest.js [hunk 3]
```diff
@@ -35,6 +43,40 @@ describe('provider testing', () => {
...
+          }).text()
+
+          const { status, statuscode } = parseXMLResponseBody(response)
+
+          if (status.value === 'ok' && statuscode.value === '100') {
+            return Promise.resolve({ description: `User '${params.userName}' added` })
+          } else if (statuscode.value === '102') {
+            return Promise.resolve({ description: `User '${params.userName}' exists` })
+          } else {
+            return Promise.reject(new Error({ description: `Cannot create user '${params.userName}'` }))
```
Comment (saw-jan) [inline tests/providerTest.js:64]: whenever error occurs, promise is rejected with error
```
(node:88758) UnhandledPromiseRejectionWarning: Error: <meaage>
```

File: tests/providerTest.js [hunk 4]
```diff
@@ -60,6 +102,47 @@ describe('provider testing', () => {
...
+              ...applicationFormUrlEncoded
+            }
+          }).text()
+
+          const { status } = parseXMLResponseBody(response)
+
+          if (status.value === 'ok') {
+            return Promise.resolve({ description: `user '${params.userName}' added to group '${params.groupName}'` })
+          } else {
+            console.log('here')
```
Comment (saw-jan) [inline tests/providerTest.js:142]:
```suggestion
            
```
remove console.log

File: tests/providerTest.js (hunk 3 above)
Comment (saw-jan) [reply tests/providerTest.js]: So, I implemented `throw new Error()`

File: tests/providerTest.js [hunk 5]
```diff
@@ -43,6 +51,42 @@ describe('provider testing', () => {
...
+          const { status, statuscode, message } = parseXMLResponseBody(response)
+
+          if (status.value === 'ok' && statuscode.value === '100') {
+            return Promise.resolve({ description: `User '${params.userName}' added` })
+          } else if (statuscode.value === '102') {
+            return Promise.resolve({ description: `User '${params.userName}' exists` })
+          } else {
+            const error = `Status: ${status.value}\nStatusCode: ${statuscode.value}\nMessage: ${message.value}`
+            throw new Error(`"Cannot create user '${params.userName}'"\n${error}`)
+          }
```
Comment (saw-jan) [inline tests/providerTest.js:74]: Anyone review this

**Comment by saw-jan** [comment]
Note: After PR https://github.com/owncloud/owncloud-sdk/pull/720/files

**Review by jasson99 — COMMENTED** [review]
LGTM. Please rebase once.

**Comment by Talank** [comment]
There are some conflicts. Please rebase with the latest master.

File: tests/groupTest.js [hunk 6]
```diff
@@ -35,8 +35,8 @@ describe('Main: Currently testing group management,', function () {
           })
             .appendElement('data', '', (data) => {
               data.appendElement('groups', '', (groups) => {
-                groups.appendElement('element', '', 'admin')
-                groups.appendElement('element', '', config.testGroup)
+                groups.appendElement('element', '', MatchersV3.string('admin'))
```
Comment (individual-it) [inline tests/groupTest.js:38]: lets use value from config here also

File: tests/providerTest.js [hunk 7]
```diff
@@ -144,6 +146,36 @@ describe('provider testing', () => {
       },
       'provider base url is returned': () => {
         return { providerBaseURL: process.env.PROVIDER_BASE_URL }
+      },
+      'user is made group subadmin': (setup, parameters) => {
+        if (setup) {
+          const response = fetch(process.env.PROVIDER_BASE_URL + '/ocs/v1.php/cloud/users/' + parameters.username + '/subadmins?format=json', {
```
Comment (individual-it) [inline tests/providerTest.js:152]: long line

File: tests/userTest.js [hunk 8]
```diff
@@ -16,7 +16,11 @@ describe('Main: Currently testing user management,', function () {
   } = require('./pactHelper.js')
   const { validAdminAuthHeaders, xmlResponseHeaders, applicationFormUrlEncoded } = require('./pactHelper.js')
 
-  const getUserInformationInteraction = function (provider, requestName, username, responseBody) {
+  const getUserInformationInteraction = async function (provider, requestName, username, responseBody) {
+    if (username === config.testUser) {
+      await provider
+        .given('the user is recreated', { username: username, password: config.testUserPassword })
+    }
```
Comment (individual-it) [inline tests/userTest.js:23]: let's check here if its NOT admin, currently we can only use `config.testUser` the given step will not work with any other user

File: tests/userTest.js [hunk 9]
```diff
@@ -34,7 +38,11 @@ describe('Main: Currently testing user management,', function () {
       })
   }
 
-  const getUsersInteraction = function (provider, requestName, query, bodyData) {
+  const getUsersInteraction = async function (provider, requestName, query, bodyData) {
+    if (Object.keys(query).length) {
```
Comment (individual-it) [inline tests/userTest.js:42]: why are we depending on the size of the query to create user or not?

File: tests/userTest.js [hunk 10]
```diff
@@ -57,9 +65,18 @@ describe('Main: Currently testing user management,', function () {
       })
   }
 
-  const changeUserAttributeInteraction = function (provider, requestName, username, requestBody, response) {
+  const changeUserAttributeInteraction = async function (provider, requestName, username, requestBody, response) {
+    if (username === config.testUser) {
+      await provider
+        .given('the user is recreated', { username: username, password: config.testUserPassword })
```
Comment (individual-it) [inline tests/userTest.js:71]: same as above, I would suggest to recreate any user that is not admin or notexisting user

File: tests/userTest.js [hunk 11]
```diff
@@ -638,8 +697,8 @@ describe('Main: Currently testing user management,', function () {
       return oc.users.setUserAttribute(config.nonExistentUser, 'email', 'asd@a.com').then(status => {
         expect(status).toBe(null)
       }).catch(error => {
-        expect(typeof (error)).toBe('object')
-        expect(error.ocs.meta.statuscode).toEqual('997')
+        expect(typeof (error)).toBe('string')
+        expect(error).toEqual('Unauthorized')
```
Comment (individual-it) [inline tests/userTest.js:701]: why do we need this change?

File: tests/userTest.js [hunk 12]
```diff
@@ -687,7 +746,7 @@ describe('Main: Currently testing user management,', function () {
         expect(status).toBe(null)
       }).catch(error => {
         expect(typeof (error)).toBe('object')
-        expect(error.ocs.meta.statuscode).toEqual('102')
+        expect(error.ocs.meta.statuscode).toEqual('103')
```
Comment (individual-it) [inline tests/userTest.js:749]: that changes the test, what is the correct status code in this case?

File: tests/userTest.js [hunk 13]
```diff
@@ -903,7 +962,7 @@ describe('Main: Currently testing user management,', function () {
           expect(status).toBe(null)
         }).catch(error => {
           expect(typeof (error)).toBe('object')
-          expect(error.ocs.meta.statuscode).toEqual('102')
+          expect(error.ocs.meta.statuscode).toEqual('103')
```
Comment (individual-it) [inline tests/userTest.js:965]: same as above

File: tests/userTest.js [hunk 14]
```diff
@@ -302,7 +362,7 @@ describe('Main: Currently testing user management,', function () {
             ocs.appendElement('meta', '', (meta) => {
               return ocsMeta(meta, 'ok', '100')
             }).appendElement('data', '', data => {
-              data.appendElement('element', '', config.testGroup)
+              data.appendElement('element', '', MatchersV3.string(config.testGroup))
```
Comment (individual-it) [inline tests/userTest.js:365]: why do we need this matcher? the group name should be the same on the mock server and on the real server, so we should be able to test it 1:1
similar for a lot of other matchers

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/userTest.js (hunk 12 above)
Comment (saw-jan) [reply tests/userTest.js:749]: I have tested manually, the status code is `103` when trying to add a non-existing user to an existing group
Also, the documentation specifies it https://doc.owncloud.com/server/10.5/developer_manual/core/apis/provisioning-api.html#add-to-group

File: tests/userTest.js (hunk 13 above)
Comment (saw-jan) [reply tests/userTest.js:965]: I have tested manually, the status code is `103` when trying to remove a non-existing user from an existing group
Also, the documentation specifies it https://doc.owncloud.com/server/10.5/developer_manual/core/apis/provisioning-api.html#add-to-group

File: tests/userTest.js (hunk 11 above)
Comment (saw-jan) [reply tests/userTest.js:701]: Yeah, It is a bit of surprising, as `setUserAttribute` is expected to return error object like in other tests but it actually returns string `Unauthorized` in case of non-existing-User

File: tests/userTest.js [hunk 15]
```diff
@@ -34,8 +38,9 @@ describe('Main: Currently testing user management,', function () {
       })
   }
 
-  const getUsersInteraction = function (provider, requestName, query, bodyData) {
+  const getUsersInteraction = async function (provider, requestName, query, bodyData) {
```
Comment (dpakach) [inline tests/userTest.js:41]: this change is not required

File: tests/userTest.js [hunk 16]
```diff
@@ -548,42 +604,19 @@ describe('Main: Currently testing user management,', function () {
...
-    )
     return provider.executeTest(async () => {
       const oc = createOwncloud()
       await oc.login()
       return oc.users.setUserAttribute(config.testUser, 'email', 'asd@a.com').then(data => {
         expect(data).toEqual(true)
-        return oc.users.getUser(config.testUser)
-      }).then(user => {
-        expect(typeof (user)).toEqual('object')
-        expect(user.email).toEqual('asd@a.com')
```
Comment (dpakach) [inline tests/userTest.js:586]: why did you delete this part?

Comment (saw-jan) [reply tests/userTest.js:586]: This part is getting user after some attribute is set. But this kind of chaining doesn't work in provider side testing (provider side testing is limited to single interaction). So, the test will fail in provider side and also the test is for testing `setUserAttributes`.

File: tests/userTest.js (hunk 11 above)
Comment (saw-jan) [reply tests/userTest.js:675]: https://github.com/owncloud/owncloud-sdk/blob/536e64836d4b51e1fa359943582b3d7a66a636f6/tests/userTest.js#L65-L74
The error is string `unauthorized` because the status code in interaction is `401` (returned by oc server when performing this request). But if the status code in interaction is set to `200`, the error returned is ocs object
Provider side testing will fail with statuscode `200` as it gets `401`

Comment (saw-jan) [reply tests/userTest.js:675]: `xyz` is non-existing user
```bash
curl -u admin:admin -X PUT http://localhost/core/ocs/v2.php/cloud/users/xyz -d key="email" -d value="example@mail.com" -v
```
```xml
* Server auth using Basic with user 'admin'
> PUT /core/ocs/v2.php/cloud/users/xyz HTTP/1.1
> Host: localhost
> Authorization: Basic YWRtaW46YWRtaW4=
> User-Agent: curl/7.68.0
> Accept: */*
> Content-Length: 32
> Content-Type: application/x-www-form-urlencoded
> 
< HTTP/1.1 401 Unauthorized
< Date: Wed, 17 Feb 2021 04:14:06 GMT
< Server: Apache/2.4.41 (Ubuntu)
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 0
< X-Robots-Tag: none
< X-Frame-Options: SAMEORIGIN
< X-Download-Options: noopen
< X-Permitted-Cross-Domain-Policies: none
< Set-Cookie: ocd3fn3x61nb=ebg5b0c8ve9j31n1qt6mo834g0; path=/core; HttpOnly
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
< Set-Cookie: oc_sessionPassphrase=lU7XJYjUsaKSuVWHNz30NSViBI2OdHsCD1qOetX58XDk1KyLGoXbnBj%2Ft4fIR0CaSrf50CmF2VpIV5yMjWeSdHZbSwXIajKlxkCuB3IWp%2BXhRyZA51mQ0wQhhi8hXVnW; path=/core; HttpOnly; SameSite=strict
< Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src * data: blob:; font-src 'self' data:; media-src *; connect-src *
< Set-Cookie: ocd3fn3x61nb=ln27kg3uh2emka932nj5fjjcq0; path=/core; HttpOnly
< Set-Cookie: cookie_test=test; expires=Wed, 17-Feb-2021 05:14:06 GMT; Max-Age=3600
< Set-Cookie: oc_username=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core; HttpOnly
< Set-Cookie: oc_token=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core; HttpOnly
< Set-Cookie: oc_remember_login=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core; HttpOnly
< Set-Cookie: oc_username=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core/; HttpOnly
< Set-Cookie: oc_token=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core/; HttpOnly
< Set-Cookie: oc_remember_login=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core/; HttpOnly
< Set-Cookie: ocd3fn3x61nb=bjbnlskdu92kg7ndvurts6gfcp; path=/core; HttpOnly
< Set-Cookie: ocd3fn3x61nb=7dvt62k1j6f73mpccdo4levnq8; path=/core; HttpOnly
* Authentication problem. Ignoring this.
< WWW-Authenticate: Basic realm="Authorisation Required"
< Content-Length: 132
< Content-Type: text/xml; charset=UTF-8
< 
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>997</statuscode>
  <message/>
 </meta>
 <data/>
</ocs>
```

Comment (saw-jan) [reply tests/userTest.js]: provider's `willRespondWith` doesn't respond provided `body` when statucode is `401`

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/userTest.js [hunk 17]
```diff
@@ -85,9 +99,22 @@ describe('Main: Currently testing user management,', function () {
...
+  const addUserToGroupInteraction = async function (provider, requestName, username, group) {
+    let ocsStatusCode = 102
+    if (username === config.nonExistentUser) {
+      ocsStatusCode = 103
+    }
+    if (username !== config.adminUsername && username !== config.nonExistentUser) {
+      await provider
+        .given('the user is recreated', { username: username, password: config.testUserPassword })
+    }
+    if (group === config.testGroup) {
```
[the remark this answers is not in the document]
Comment (saw-jan) [reply tests/userTest.js:111]: there's also a test with non-existing group
So, group is created with only existing group test
But adding group in all tests won't fail the interactions, we can add group in every interactions.

**Comment by saw-jan** [comment]
That passed before because the status code was `200` in `willResponsWith` earlier and consumer side will accept that. But with provider side, it expects `401`

File: tests/userTest.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (saw-jan) [reply tests/userTest.js]: https://github.com/owncloud/owncloud-sdk/blob/36b7737ae94e16c6fa7cc09aed6826afe6e98bfc/tests/userTest.js#L111-L114
I think this works better

File: tests/userTest.js (hunk 11 above)
Comment (individual-it) [reply tests/userTest.js:675]: the status code should be `200` for v1 API and `401` for v2 requests
the interaction is set to use `v1` https://github.com/owncloud/owncloud-sdk/pull/714/files#diff-b799f43eb8d74a22d0a06495f79b2b1ec471eaa171dc202b5d564fb9ebca4064R80
your curl request uses `v2`

Comment (saw-jan) [reply tests/userTest.js:675]: I also thought `v2` may be the problem, but same result with `v1` also

Comment (saw-jan) [reply tests/userTest.js:675]:
```
curl -u admin:admin -X PUT http://localhost/core/ocs/v1.php/cloud/users/xyz -d key="email" -d value="example@mail.com" -v

*   Trying 127.0.0.1:80...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 80 (#0)
* Server auth using Basic with user 'admin'
> PUT /core/ocs/v1.php/cloud/users/xyz HTTP/1.1
> Host: localhost
> Authorization: Basic YWRtaW46YWRtaW4=
> User-Agent: curl/7.68.0
> Accept: */*
> Content-Length: 32
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 32 out of 32 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Date: Wed, 17 Feb 2021 06:02:20 GMT
< Server: Apache/2.4.41 (Ubuntu)
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 0
< X-Robots-Tag: none
< X-Frame-Options: SAMEORIGIN
< X-Download-Options: noopen
< X-Permitted-Cross-Domain-Policies: none
< Set-Cookie: ocd3fn3x61nb=1hnt3f4v6derp3tma8ajpo6247; path=/core; HttpOnly
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
< Set-Cookie: oc_sessionPassphrase=9dGRD6bVRjpAenJe6RzikKSETdMNTxt9puIp%2FqmDnY12b4XnGLGiFWk%2F8KVwMSq6WOsoYw9QgGf0Ums9c7hKoxnAANWsxvwiqGFjFk4OpLOQN%2FQW0UuX%2Fm6GnX64rS27; path=/core; HttpOnly; SameSite=strict
< Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src * data: blob:; font-src 'self' data:; media-src *; connect-src *
< Set-Cookie: ocd3fn3x61nb=8t462ol86ohc3smf7k49gksrnc; path=/core; HttpOnly
< Set-Cookie: cookie_test=test; expires=Wed, 17-Feb-2021 07:02:20 GMT; Max-Age=3600
< Set-Cookie: oc_username=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core; HttpOnly
< Set-Cookie: oc_token=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core; HttpOnly
< Set-Cookie: oc_remember_login=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core; HttpOnly
< Set-Cookie: oc_username=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core/; HttpOnly
< Set-Cookie: oc_token=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core/; HttpOnly
< Set-Cookie: oc_remember_login=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/core/; HttpOnly
< Set-Cookie: ocd3fn3x61nb=pqgleurba0bt1tkedadqttuba6; path=/core; HttpOnly
< Set-Cookie: ocd3fn3x61nb=hicoltb4584aokdr581ofhc2ni; path=/core; HttpOnly
* Authentication problem. Ignoring this.
< WWW-Authenticate: Basic realm="Authorisation Required"
< Content-Length: 132
< Content-Type: text/xml; charset=UTF-8
< 
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>997</statuscode>
  <message/>
 </meta>
 <data/>
</ocs>
* Connection #0 to host localhost left intact
```

**Comment by saw-jan** [comment]
statuscode `401` and ocs `997` from server seems to be a bug https://github.com/owncloud/core/issues/38423.
I have skipped the related test (setAttribute to non-existing user) and commented the issue.

File: tests/userTest.js [hunk 18]
```diff
@@ -618,17 +647,19 @@ describe('Main: Currently testing user management,', function () {
     })
   })
 
-  it('checking method : setUserAttribute of a non existent user', async function () {
+  // TODO: unskip the test after the issue has been fixed
+  // https://github.com/owncloud/core/issues/38423
+  it.skip('checking method : setUserAttribute of a non existent user', async function () {
```
Comment (individual-it) [inline tests/userTest.js:652]: lets have this test running with the correct expectations on the consumer side, its not an issue if it currently fails on the provider side, that will not fail CI

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/groupTest.js [hunk 19]
```diff
@@ -35,8 +35,8 @@ describe('Main: Currently testing group management,', function () {
           })
             .appendElement('data', '', (data) => {
               data.appendElement('groups', '', (groups) => {
-                groups.appendElement('element', '', 'admin')
-                groups.appendElement('element', '', config.testGroup)
+                groups.appendElement('element', '', MatchersV3.string(config.adminUsername))
+                groups.appendElement('element', '', MatchersV3.string(config.testGroup))
```
Comment (individual-it) [inline tests/groupTest.js:39]: I don't think we need matchers here, the mock server and the provider should respond with the exact values

File: tests/userTest.js [hunk 20]
```diff
@@ -451,7 +503,7 @@ describe('Main: Currently testing user management,', function () {
       data => {
         data.appendElement('users', '', users => {
           users.appendElement('element', '', config.adminUsername)
-            .appendElement('element', '', config.testUser)
+            .appendElement('element', '', MatchersV3.string(config.testUser))
```
Comment (individual-it) [inline tests/userTest.js:506]: I don't think we need matchers here, the mock server and the provider should respond with the exact value

**Review by individual-it — CHANGES_REQUESTED** [review]

**Review by individual-it — APPROVED** [review]
